ci/eval/compare/utils.nix

```diff
@@ -0,0 +1,125 @@
+{ lib, ... }:
+rec {
+  _processSystemPath =
+    packageSystemPath:
+    let
+      # python312Packages.torch.aarch64-linux -> ["python312Packages" "torch" "aarch64-linux"]
+      # splittedPath = lib.splitString "." attrName;
+      splittedPath = lib.splitString "." packageSystemPath;
+
+      # ["python312Packages" "torch" "aarch64-linux"] -> ["python312Packages" "torch"]
+      packagePath = lib.sublist 0 (lib.length splittedPath - 1) splittedPath;
+    in
+    {
+      # "python312Packages.torch"
+      name = lib.concatStringsSep "." packagePath;
+
+      # "aarch64-linux"
+      system = lib.last splittedPath;
+    };
+
+  # Turns
+  # [
+  #   "hello.aarch64-linux"
+  #   "hello.x86_64-linux"
+  #   "hello.aarch64-darwin"
+  #   "hello.x86_64-darwin"
+  #   "bye.x86_64-darwin"
+  #   "bye.aarch64-darwin"
+  # ]
+  #
+  # into
+  #
+  # [
+  #   "hello"
+  #   "bye"
+  # ]
+  extractPackageNames =
+    packageSystemPaths:
+    lib.unique (
+      lib.filter (n: n != "") (map (attrName: (_processSystemPath attrName).name) packageSystemPaths)
+    );
+
+  # Computes a diff between two attrs
+  # {
+  #   added: [ <keys only in the second object> ],
+  #   removed: [ <keys only in the first object> ],
+  #   changed: [ <keys with different values between the two objects> ],
+  # }
+  #
+  diff =
+    let
+      filterKeys = cond: attrs: lib.attrNames (lib.filterAttrs cond attrs);
+    in
+    old: new: {
+      added = filterKeys (n: _: !(old ? ${n})) new;
+      removed = filterKeys (n: _: !(new ? ${n})) old;
+      changed = filterKeys (
+        n: v:
+        # Filter out attributes that don't exist anymore
+        (new ? ${n})
+
+        # Filter out attributes that are the same as the new value
+        && (v != (new.${n}))
+      ) old;
+    };
+
+  # Turns
+  # [
+  #   "hello.aarch64-linux"
+  #   "hello.x86_64-linux"
+  #   "hello.aarch64-darwin"
+  #   "hello.x86_64-darwin"
+  #   "bye.x86_64-darwin"
+  #   "bye.aarch64-darwin"
+  # ]
+  #
+  # into
+  #
+  # {
+  #   linux = [
+  #     "hello"
+  #   ];
+  #   darwin = [
+  #     "hello"
+  #     "bye"
+  #   ];
+  # }
+  groupByKernel =
+    systemPaths:
+    let
+      systemPaths' = builtins.map _processSystemPath systemPaths;
+
+      filterKernel =
+        kernel:
+        lib.unique (
```

Comment:
Btw. Here is a faster way to compute unique string sets: https://github.com/NixOS/nixpkgs/pull/355616/files#diff-747a1e19dd64dc3c078800ad4bbcc30f4368af5c562960ac2437049684b694dbR1831

Reply:
Untested, but hopefully clear how it works:

```
diff --git ci/eval/compare/utils.nix ci/eval/compare/utils.nix
index 5f7e39262215..6d9f935a6ac4 100644
--- ci/eval/compare/utils.nix
+++ ci/eval/compare/utils.nix
@@ -92,8 +92,8 @@ rec {
 
       filterKernel =
         kernel:
-        lib.unique (
-          lib.map (systemPath: systemPath.name) (
+        builtins.attrNames (
+          builtins.groupBy (systemPath: systemPath.name) (
             builtins.filter (systemPath: lib.hasSuffix kernel systemPath.system) systemPaths'
           )
         );
```

Reply:
lib.unique is O(n^2)

Reply:
Ah. actually this is not a string set...

Reply:
```
diff --git ci/eval/compare/utils.nix ci/eval/compare/utils.nix
index 5f7e39262215..f82df8fa8650 100644
--- ci/eval/compare/utils.nix
+++ ci/eval/compare/utils.nix
@@ -92,13 +92,14 @@ rec {
 
       filterKernel =
         kernel:
-        lib.unique (
-          lib.map (systemPath: systemPath.name) (
-            builtins.filter (systemPath: lib.hasSuffix kernel systemPath.system) systemPaths'
-          )
+        builtins.groupBy (systemPath: systemPath.name) (
+          builtins.filter (systemPath: lib.hasSuffix kernel systemPath.system) systemPaths'
         );
     in
-    lib.genAttrs [ "linux" "darwin" ] filterKernel;
+      {
+        linux = filterKernel "linux";
+        darwin = filterKernel "darwin";
+      };
 
   getLabels = lib.mapAttrs (
     kernel: rebuildCount:
```

Reply:
```patch
diff --git ci/eval/compare/utils.nix ci/eval/compare/utils.nix
index 5f7e39262215..6d260e5d34b3 100644
--- ci/eval/compare/utils.nix
+++ ci/eval/compare/utils.nix
@@ -36,8 +36,8 @@ rec {
   # ]
   extractPackageNames =
     packageSystemPaths:
-    lib.unique (
-      lib.filter (n: n != "") (map (attrName: (_processSystemPath attrName).name) packageSystemPaths)
+    builtins.attrNames (
+      builtins.removeAttrs (builtins.groupBy (packageSystemPath: (_processSystemPath packageSystemPath).name) packageSystemPaths) [""]
     );
 
   # Computes a diff between two attrs
@@ -92,8 +92,8 @@ rec {
 
       filterKernel =
         kernel:
-        lib.unique (
-          lib.map (systemPath: systemPath.name) (
+        builtins.attrNames (
+          builtins.groupBy (systemPath: systemPath.name) (
             builtins.filter (systemPath: lib.hasSuffix kernel systemPath.system) systemPaths'
           )
         );
```

Reply:
Thanks for the tips. I made the proposed changes.